src/util/bitvector_expr.h

```diff
@@ -778,6 +781,66 @@ inline binary_overflow_exprt &to_binary_overflow_expr(exprt &expr)
   return ret;
 }
 
+class plus_overflow_exprt : public binary_overflow_exprt
+{
+public:
+  plus_overflow_exprt(exprt _lhs, const irep_idt &kind, exprt _rhs)
+    : binary_overflow_exprt(std::move(_lhs), ID_overflow_plus, std::move(_rhs))
+  {
+  }
+};
+
+template <>
+inline bool can_cast_expr<plus_overflow_exprt>(const exprt &base)
+{
+  return base.id() == ID_overflow_plus;
+}
+
+class minus_overflow_exprt : public binary_overflow_exprt
+{
+public:
+  minus_overflow_exprt(exprt _lhs, const irep_idt &kind, exprt _rhs)
+    : binary_overflow_exprt(std::move(_lhs), ID_overflow_minus, std::move(_rhs))
+  {
+  }
+};
+
+template <>
+inline bool can_cast_expr<minus_overflow_exprt>(const exprt &base)
+{
+  return base.id() == ID_overflow_minus;
+}
+
+class mult_overflow_exprt : public binary_overflow_exprt
+{
+public:
+  mult_overflow_exprt(exprt _lhs, const irep_idt &kind, exprt _rhs)
+    : binary_overflow_exprt(std::move(_lhs), ID_overflow_mult, std::move(_rhs))
+  {
+  }
+};
+
+template <>
+inline bool can_cast_expr<mult_overflow_exprt>(const exprt &base)
+{
+  return base.id() == ID_overflow_mult;
+}
+
+class shl_overflow_exprt : public binary_overflow_exprt
+{
+public:
+  shl_overflow_exprt(exprt _lhs, const irep_idt &kind, exprt _rhs)
+    : binary_overflow_exprt(std::move(_lhs), ID_overflow_shl, std::move(_rhs))
+  {
+  }
+};
+
+template <>
+inline bool can_cast_expr<shl_overflow_exprt>(const exprt &base)
+{
+  return base.id() == ID_overflow_shl;
+}
+
```

Comment:
Thank you for adding these. But then these are missing some `unary_minus_overflow_exprt` (which is a `unary_overflow_exprt`).

Reply:
Thank-you for the reminder. This could do with adding for completeness. However I have found that C examples with unary minus can be handled by cbmc being run with the new `--incremental-smt2-solver` argument, without supporting them in the new decision procedure, due to the checks being expressed in terms of more fundamental bit vector expressions in `goto_check_c` - https://github.com/diffblue/cbmc/blob/a50ef65bc050f1122fca10d55ad5c8622d8f679a/src/ansi-c/goto_check_c.cpp#L859

As I read the code in `goto_check_c` and the code in the decision procedure, it seems that the implementation of the overflow checks is split / duplicated between these two different places. Do you know if there are any plans to change how this is structured?